predict_with_ensemble

AviralVijay-GSLab opened the pull request:
@michaellevy  I have created the separate branch for ensemble capabilities in predict.R file. I have used the average out method for ensembling because find a few facts here:-
1.  Random forest reduces variance and boosting reduces bias on the other hand XGboost have capabilities to reduce variance & bias, therefore, XGboost should be best selected model in that case.
2. Random forest was the best model among the all three based on the diabetes data, therefore its performance matrix based on the data provided by user which can be varied, in such condition, I have provided equal weight to all three models( i.e. +1) so that ensemble prediction will have ability to the best performing models also & the other ones also which can contribute to reduce variance or bias or both.

Please suggest if you think any other way which can fit best here apart from this one.

michaellevy commented:
I like the approach you've taken here, @AviralVijay-GSLab. A couple of main points:

1. We need a way to increase the contribution of the best-performing model to the ensemble prediction. With equal weight as you've done, the ensemble could easily under-perform the best individual model. I just looked quickly, but I think think [this article](https://medium.com/weightsandbiases/an-introduction-to-model-ensembling-63effc2ca4b3) discusses some ways to do that.

2. There is a lot of repeated code both within and across `predict_model_list_main` and `get_oof_predictions`. As a general programming principle, we avoid repeated code to make it easier to understand and change. Putting the repeated code into a function that is called by those functions would be better.

3. As written, the code assumes there will be three models in the model_list, but this might not be the case (e.g. `machine_learn(mtcars, outcome = mpg, models = c("rf", "xgb"))`, or if we add additional algorithms to the package). In R, when you're doing something over each object in a list, `lapply` or one of the `purrr::map_` functions are preferable to extracting and using each item individually. This will help with both reduce repeated code and generalizability.

AviralVijay-GSLab commented:
Hi @michaellevy  , Hope all good :)
I have completed those changes along with the weighted average method for ensemble predictions where best_performing_model always have maximum weight i.e +1, Please have a look at it & suggest if you want to suggest more changes in it.

michaellevy commented:
Hi @AviralVijay-GSLab thanks for your continued? work on this. Can you explain to me how you're weighting the models and what this glm is doing? `weight_modeling <- glm(target ~ ., data = preds_data_frame)`

AviralVijay-GSLab commented:
Sure @michaellevy , The sole purpose of applying glmnet or Generalized linear model here is to find the coefficient of features respectively which can be used to calculate the weight to be assigned to the individual model latter in the same function using add method after multiplying with their weights.

Classification classifier works on below equation where  ŷi is used as response variable.

Model: ŷi = sign(Score(xi))
Score(xi) = w0 + w1 xi[1] + ... + wd xi[d] = wTxi
feature 1 = 1
feature 2 = x[1] ...eg. in our case:  model 1[first]
feature 3 = x[2] ...eg. in our case model  2[second]
feature d+1 = x[d] ...eg. in our case model d

We only want to find the wT(weights or coefficients) of features using glmnet.
for more details, I have attached one pdf while in debug mode for insight scoop of function level details.

Please connect if need more details.

[Explaination_weight_models.pdf](https://github.com/HealthCatalyst/healthcareai-r/files/2293227/Explaination_weight_models.pdf)
.

michaellevy commented:
@AviralVijay-GSLab, do I have this right: You're fitting a model with the best-model's predictions as the response variable and the other models' predictions as the predictors? If this is an established method, please provide references. If not we'd need extensive testing that this improves performance across a variety of datasets and prediction tasks. 

If we were to take this approach, I'm also not sure why we wouldn't use the training data response variable as the outcome and the out of fold predictions as the predictors. I'm also worried about multicollinearity among the models leading to a biased assessment of how important each model is. 

Finally, the code only seems to work when `newdata` is not provided to `predict`, and in that scenario, performance is unreasonably good for classification problems. See demo below. We'd need extensive unit and integration tests to ensure that the functionality works as intended.

``` r
library(healthcareai)
#> healthcareai version 2.1.1
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com
# Classification
d <- split_train_test(pima_diabetes, diabetes, seed = 265)
set.seed(54729)
m <- machine_learn(d$train, patient_id, outcome = diabetes, tune = FALSE)
#> Training new data prep recipe...
#> Variable(s) ignored in prep_data won't be used to tune models: patient_id
#> 
#> diabetes looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 12 features with 615 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 5 rf's, 5 xgb's, and 50 glm's
#> Training at fixed values: Random Forest
#> Training at fixed values: eXtreme Gradient Boosting
#> Training at fixed values: glmnet
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***

# Predict on test dataset: ensemble doesn't do anything
preds_ens <- predict(m, d$test, ensemble = TRUE)
#> Prepping data based on provided recipe
preds_one <- predict(m, d$test)
#> Prepping data based on provided recipe
all.equal(evaluate(preds_ens), evaluate(preds_one))
#> [1] TRUE

# Predict on training dataset: Predictions are too good
oof_preds_ens <- predict(m, ensemble = TRUE)
oof_preds_one <- predict(m)
evaluate(oof_preds_ens)
#>      AUPR     AUROC 
#> 0.9953488 1.0000000
evaluate(oof_preds_one)
#>      AUPR     AUROC 
#> 0.7020599 0.8310698


# Regression
d <- split_train_test(pima_diabetes, age, seed = 265)
set.seed(54729)
m <- machine_learn(d$train, patient_id, outcome = age, tune = FALSE)
#> Training new data prep recipe...
#> Variable(s) ignored in prep_data won't be used to tune models: patient_id
#> 
#> age looks numeric, so training regression algorithms.
#> 
#> After data processing, models are being trained on 14 features with 616 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 5 rf's, 5 xgb's, and 50 glm's
#> Training at fixed values: Random Forest
#> Training at fixed values: eXtreme Gradient Boosting
#> Training at fixed values: glmnet
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***

# Predict on test dataset: ensemble still doesn't do anything
preds_ens <- predict(m, d$test, ensemble = TRUE)
#> Prepping data based on provided recipe
preds_one <- predict(m, d$test)
#> Prepping data based on provided recipe
all.equal(evaluate(preds_ens), evaluate(preds_one))
#> [1] TRUE

# Predict on training dataset: This several-percent improvement from ensemble looks reasonable to me
oof_preds_ens <- predict(m, ensemble = TRUE)
oof_preds_one <- predict(m)
evaluate(oof_preds_ens)
#>      RMSE       MAE  Rsquared 
#> 9.4227312 6.7619243 0.3727228
evaluate(oof_preds_one)
#>      RMSE       MAE  Rsquared 
#> 9.6469967 6.9554036 0.3427261
```

Created on 2018-08-20 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

AviralVijay-GSLab commented:
@michaellevy Please find the [reference article ](https://www.listendata.com/2015/08/ensemble-learning-stacking-blending.html) , Yes we can use the training data response variable as outcome variable, it works well for regression because predicted and actual response variable both are numerical i.e model's outcome response variable also have numeric data as training dataset actual response variable have.
But on the other hand, For classification problem, Let's take diabetes(which is training data outcome response variables) have two classes Y and N but model's predicted outcome is in numeric form i.e in the form of percentages or probability. 
In that case Predicted and Actual Variables are in different form therefore We have used the Best model's predicted variable as response variable. Kindly have a look at reference article above mentioned for the same. As you suggested We can only make this functionality without changing the package's use design therefore it was the way which can be implemented to achieve that.
In general cases, classification models usually provides predicted classes also along with probabilities, but it is not same in our case.  
to_pred is supplied as new data when using_training_data is false, Thus it depend on the data which comes  to it.
We agree to have the extensive testing of this functionality because this implementation can increase the overall prediction for the package.
Please have a look below and suggest further.

![ensemble _ new_data](https://user-images.githubusercontent.com/41198318/44401401-2627e580-a56c-11e8-8681-d4f7f5ce5f90.PNG)

michaellevy commented:
@AviralVijay-GSLab We need to put this on backburner for the next few weeks as we get ready for our the v2.2 release and our Healthcare Analytics Summit. If you want to continue to work on it in the meantime, here are a few ideas:

- By a reference, I mean a study (peer reviewed, or at minimum and arxiv paper) demonstrating that this method is effective. 

> But on the other hand, For classification problem, Let's take diabetes(which is training data outcome response variables) have two classes Y and N but model's predicted outcome is in numeric form...

- There are, of course, statistical models that support this; in fact, that's what the `glm` function is for. Take a look at its `family` argument.

- Tests. We need lots of tests, including to address the issues I raised in my previous comment.

mmastand commented:
@AviralVijay-GSLab Has @michaellevy's latest round of comments been addressed?

AviralVijay-GSLab commented:
@mmastand, @michaellevy raised basically two issue along with extensive unit and integration testing.
1. training data should be used as label in place of best models predictions.
Our take: It can be done where we have the outcome labels but what if in future one want to utilise this package functionality for developing the recommendation system or in unsupervised learning scenario, our solution can also be fit in supervised and unsupervised learning arena, therefore, we have evolved this approach.
2. multicollinearity among the models leading to a biased assessment of how important each model is:
Our take: To decide the importance of models, we took best model as reference because best model is decided by the existing package functionality, we have assumed that best models predictions near to the actual values of target variable, therefore, Best model will always have the weight +1 and other models have their respective weights, to reduce the bias, we have not removed the negative coefficient models prediction, we have applied abs there which returns absolute Positive value of a coefficient, therefore all the models contribute to the final predictions based on their weights.
 We have initiated our approach to publish in the mainstream journal, will share the link once we got the decision on that, basically biased and variance can be handles via different regularisation techniques, we have evolved the core method of weighted average ensemble to make it more feasible and effective.
I think, these issues addressed well, meanwhile we suggest to let it validated with different scenarios.
@SameerMahajan-GSLab

mmastand commented:
I've taken a look at the newest version. I agree with @michaellevy's assessment, which was:
- You need to provide a reference (peer-reviewed or arxiv) that showcases the benefits of the method over a variety of scenarios.
- The code doesn't appear to add any value vs. predicting with a single model. I've tested it using a separate, larger, data set with more complete model tuning. See the reprex below, using `adult.data` from https://archive.ics.uci.edu/ml/datasets/Adult 

```
**library(healthcareai, quietly = TRUE)
#> healthcareai version 2.1.1
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com
library(tidyverse, quietly = TRUE)
#> Warning: package 'tidyverse' was built under R version 3.4.2
#> Warning: package 'tibble' was built under R version 3.4.3
#> Warning: package 'tidyr' was built under R version 3.4.4
#> Warning: package 'purrr' was built under R version 3.4.4
#> Warning: package 'dplyr' was built under R version 3.4.4
#> Warning: package 'stringr' was built under R version 3.4.4
d <- read_csv("~/Desktop/adult_train.csv", col_names = FALSE) %>%
  mutate(X15 = ifelse(X15 == ">50K", "yes", "no")) %>%  
  dplyr::na_if("?")
#> Parsed with column specification:
#> cols(
#>   X1 = col_integer(),
#>   X2 = col_character(),
#>   X3 = col_integer(),
#>   X4 = col_character(),
#>   X5 = col_integer(),
#>   X6 = col_character(),
#>   X7 = col_character(),
#>   X8 = col_character(),
#>   X9 = col_character(),
#>   X10 = col_character(),
#>   X11 = col_integer(),
#>   X12 = col_integer(),
#>   X13 = col_integer(),
#>   X14 = col_character(),
#>   X15 = col_character()
#> )
#> Warning: package 'bindrcpp' was built under R version 3.4.4

d <- split_train_test(d, X15)

m <- d$train %>%
  machine_learn(X14, outcome = X15, tune = TRUE, tune_depth = 5)
#> Warning in prep_data(d, !(!(!dots)), outcome = !(!outcome), impute =
#> impute): These ignored variables have missingness: X14
#> Training new data prep recipe...
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   X11 and X12
#> Variable(s) ignored in prep_data won't be used to tune models: X14
#> 
#> X15 looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 49 features with 26,049 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 25 rf's, 25 xgb's, and 50 glm's 
#> Model training may take a few minutes.
#> Training with cross validation: Random Forest
#> Warning: package 'caret' was built under R version 3.4.3
#> Training with cross validation: eXtreme Gradient Boosting
#> Training with cross validation: glmnet
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***


p <- predict(m, newdata = d$train)
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   X11 and X12
pe <- predict(m, newdata = d$train, ensemble = TRUE)
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   X11 and X12
all.equal(evaluate(p), evaluate(pe))
#> [1] "Mean relative difference: 1.02906e-06"

evaluate(p)
#>      AUPR     AUROC 
#> 0.7781745 0.9165491
evaluate(pe)
#>      AUPR     AUROC 
#> 0.7781761 0.9165492

p_test <- predict(m, newdata = d$test)
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   X11 and X12
pe_test <- predict(m, newdata = d$test, ensemble = TRUE)
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   X11 and X12
all.equal(evaluate(p_test), evaluate(pe_test))
#> [1] "Mean relative difference: 1.720494e-07"

evaluate(p_test)
#>     AUPR    AUROC 
#> 0.711133 0.890783
evaluate(pe_test)
#>      AUPR     AUROC 
#> 0.7111329 0.8907828
```

If you plan to use your own Arxiv paper as the reference to show the validity of the method, we'd like to wait until it's accepted and published before dealing with code style, unit tests, or conflicts. Additionally, like to see improvement on established methods across a variety of data sets. If you can demonstrate better accuracy, please do.

AviralVijay-GSLab commented:
Hi @mmastand , We have sent our paper to the Journal for publication, it takes about 8-12 weeks(already 3 weeks completed). We will let you know once this would be published. 

Meanwhile, Please find the attached "Results_in_Nutshell.pdf" file for significant improvement(42% to 63% and 42% to 97%) in accuracy while using 'ensemble=TRUE' in predict.R function on training data.
We have used below dataset available on UCI for these results:

1. [Breast Cancer Data Set](https://archive.ics.uci.edu/ml/datasets/Breast+Cancer)
2. [EEG Eye State Data Set](https://archive.ics.uci.edu/ml/datasets/EEG+Eye+State) 
3. [Haberman's Survival Data Set ](https://archive.ics.uci.edu/ml/datasets/Haberman%27s+Survival)

Although, currently it's not processing the newdata, therefore provide the same accuracy on testSet, we are working to figure out the reason of it because we have passed the required parameters in that.

[Results_in_Nutshell.pdf](https://github.com/HealthCatalyst/healthcareai-r/files/2542284/Results_in_Nutshell.pdf)

Please connect, if require more information.


@sameermahajan

mmastand commented:
Hi @sameermahajan,
Congrats on the paper. Do you know why the test set accuracy is the same as without ensemble? Is it a problem with your code? Or does the algorithm not extend to new data?

AviralVijay-GSLab commented:
Hi @mmastand 

I have fixed the isssue. Now Ensemble works on new data also. Please find the the below Results in Nutshell.pdf.
We have achieved significant accuracy improvement on test data like 9%, 22%. 
Please suggest if need anymore information from our end.
[Results_in_Nutshell.pdf](https://github.com/HealthCatalyst/healthcareai-r/files/2604356/Results_in_Nutshell.pdf)

@sameermahajan

AviralVijay-GSLab commented:
Hi @mmastand,

Hope you're great. Just sending you a gentle reminder about above solution for ensemble capabilities.
We would like to hear from your end.

@sameermahajan

mmastand commented:
@AviralVijay-GSLab  Working on this now, should see something today or tomorrow.

glenrs commented:
@AviralVijay-GSLab I have not been involved. I was asked to give my opinion in addition to the others.

`healthcareai` was written to make machine learning effective and easily interpretable. We have chosen to use the current models for these reasons. We could have included many other peer reviewed and widely accepted algorithms that would improve performance slightly. We have not chosen to do this because performance improvement does not out weigh the cost in interpretability.

Having said this, there are many other things that concern me about this implementation. Your results do not include any plots. Sometimes metrics can be misleading. In the examples below the metrics improve, but I would much prefer the previous implementation because of the more separated distributions. Even though metrics improve doesn't mean performance does!

The first graph doesn't even contain any "Y". There is a problem somewhere in the code somewhere. This is a big bug.

There are not very many tests for the functionality in this PR. Consider making more unit tests for other PRs in the future.

I apologize for the work you have put into this, but in my opinion it should not be added to healthcareai because it doesn't improve performance enough to outweigh interpretability. We have not used neural networks and other methods for similar reasons, and they are peer reviewed!

``` r
library(healthcareai)
#> healthcareai version 2.1.1
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

d <-
  nycflights13::flights %>%
  mutate(carrier = ifelse(carrier %in% c("B6", "UA"), "Y", "N")) %>%
  slice(1:2000)

d %>%
  pull(carrier) %>%
  table()
#> .
#>    N    Y 
#> 1262  738

d <- split_train_test(d, carrier, seed = 265)
set.seed(54729)
m <- machine_learn(d$train, outcome = carrier, tune = FALSE)
#> Training new data prep recipe...
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   year and month
#> 
#> carrier looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 36 features with 1,601 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 5 rf's, 5 xgb's, and 50 glm's
#> Training at fixed values: Random Forest
#> Training at fixed values: eXtreme Gradient Boosting
#> Training at fixed values: glmnet
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***

suppressWarnings(preds_ens <- predict(m, d$test, ensemble = TRUE))
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   year and month
suppressWarnings(preds_one <- predict(m, d$test))
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   year and month
all.equal(evaluate(preds_ens), evaluate(preds_one))
#> [1] "Mean relative difference: 0.008294937"

suppressWarnings(oof_preds_ens <- predict(m, ensemble = TRUE))
suppressWarnings(oof_preds_one <- predict(m))
evaluate(oof_preds_ens)
#>     AUPR    AUROC 
#> 0.998308 1.000000
evaluate(oof_preds_one)
#>      AUPR     AUROC 
#> 0.9464988 0.9763482

# This shouldn't happen. This is very wrong. 
suppressWarnings(predict(m, ensemble = TRUE)) %>% 
    plot()
#> Warning: Removed 599 rows containing non-finite values (stat_density).
```

![](https://i.imgur.com/SptPfQJ.png)

``` r
  
suppressWarnings(predict(m)) %>% 
    plot()
```

![](https://i.imgur.com/j4iEwPO.png)

``` r
  
suppressWarnings(predict(m, d$test, ensemble = TRUE)) %>%
    plot()
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   year and month
#> Warning: Removed 143 rows containing non-finite values (stat_density).
```

![](https://i.imgur.com/AKAmCNC.png)

``` r
  
suppressWarnings(predict(m, d$test)) %>%
    plot()
#> Prepping data based on provided recipe
#> Removing the following 2 near-zero variance column(s). If you don't want to remove them, call prep_data with remove_near_zero_variance as a smaller numeric or FALSE.
#>   year and month
```

![](https://i.imgur.com/CM9rtoy.png)

Created on 2018-12-06 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

AviralVijay-GSLab commented:
Hi @glenrs ,

Thanks for connecting, I would like you to know about the previous context you might not be known with.
Please have a look at the closed PR i.e [Ensemble Function added here](https://github.com/HealthCatalyst/healthcareai-r/pull/1205). #1205 
I would like to answer your questions by referring a few conversation with @michaellevy .

1.  Kindly refer his third comment (second paragraph)  in the pr #1205, where it is clearly mentioned that "but it has to be done in a way that fits the package's use design." , therefore, we have implemented the Ensemble in a way that it will not impact the existing package design, here line no.170 in predict function, you can see that preds only have the predictions of positive class of best model. therefore, we have implemented the ensemble only on positive class of predictions, Thus it was not the defect, we have implemented it in the way to fit into the package use design( Kindly refer the attached PDF for same).
2.  We have not predicted anything, we have only use the predictions of the existing models in the same way it was in the predict function(preds[[mi$positive_class]], assuming that existing predict function align with the package use design.
3. We strongly recommend the interpretation of Evaluation matrix, therefore, We have picked the confusion matrix in place of evalution function inside the healthcare package because, confusion matrix provide the clear interpretation of TRUE Positive, FALSE Positive, FALSE Negative and TRUE Negative, but we have not used it due to fit into the package use design.
4. Interpretation can be done by using Data Visualisation and Caret have rich implementation of Evaluation matrix like Adjusted R Squared, Confusion matrix(Kindly refer the fourth comment of him in the PR #1205 where it is mentioned that "You'll see that predict.model_list only returns probabilities for classification problems, "), therefore, we have focused on ensemble capabilities assuming that package use design have enough capabilities for interpretation if evalution function is used in place of confusion matrix.
5. We believe to work in an Agile environment, therefore, If you suggest to have the predictions on positive and negative classes then we can modify our existing implementation.
6. We accept that Testing is required but its same as the most of the outstanding solutions.

Please find the attached pdf.

@mmastand @sameermahajan

Please suggest further.

[Reference.pdf](https://github.com/HealthCatalyst/healthcareai-r/files/2656772/Reference.pdf)

mmastand commented:
Thank you everyone for the discussion. Thank you for taking the time to fix your code and integrate it into our current framework. 

I've had a chance to thoroughly evaluate the added functionality. At this point, my main concern is that the method is overfitting on training data and not improving performance on new data. I've used 4 different data sets to assess this and haven't seen any benefits. 

I noticed that the "results in nutshell" that you provided showed improvement in a couple of cases where the data was only a few hundred rows. I think we can agree that data is typically larger in real world scenarios.

Finally, as @glenrs mentioned, I'm worried about interpretability. Variable importance or coefficient weighting based explanations are very important to our users, and those are lost when the final model is being trained on other models. While it's a very interesting and novel approach, I don't think it is appropriate for our package. 

Please see the attached pdf for my evaluation of the functionality.  
[ensemble_pr.pdf](https://github.com/HealthCatalyst/healthcareai-r/files/2657911/ensemble_pr.pdf)

AviralVijay-GSLab commented:
Thanks, @mmastand  for sharing your concerns.

I would like to address the overfitting issue on next week when I would able to go through it.

But I would like to clarify one point above mentioned about interpretability, i.e "when the final model is being trained on other models. "
This is one of the incorrect interpretation about our solution. I repeat, we have not trained the final model to make the predictions.
Correct Interpretation: " Ensemble computes the weights of the predictions of existing trained models using the estimate of glm, This glm is not trained as final model, i.e, only averages is being taken of the predictions based on their weights, we have not re-predict using the said final model".

Hope it useful. We have not trained any model to predict anything, we have only computes the weights.
 
I would get back to you if I would able to show the relative big dataset to address the overfitting issue.

Rest, It was nice to connect with you.... :+1: 

Thanks again ..!

mmastand commented:
Congratulations on your package! It looks interesting. I think the method is more appropriate there. Thank you for the discussion.
